javalib/src/main/scala/java/math/BigInteger.scala

```diff
@@ -0,0 +1,1500 @@
+   */
+  override def doubleValue(): Double = {
+    java.lang.Double.parseDouble(this.toString)
+  }
+
+  /**
+   * Returns {@code true} if {@code x} is a BigInteger instance and if this
+   * instance is equal to this {@code BigInteger}.
+   *
+   * @param x object to be compared with {@code this}.
+   * @return true if {@code x} is a BigInteger and {@code this == x}, {@code
+   *         false} otherwise.
+   */
+  override def equals(x: Any): Boolean = x match{
+    case that: BigInteger => this.sign == that.sign && this.numberLength == that.numberLength && this.equalsArrays(that.digits)
+    case _ => false
+  }
+
+
+  /**
+   * Returns a new {@code BigInteger} which has the same binary representation
+   * as {@code this} but with the bit at position n flipped. The result is
+   * equivalent to {@code this ^ 2^n}.
+   * <p>
+   * <b>Implementation Note:</b> Usage of this method is not recommended as the
+   * current implementation is not efficient.
+   *
+   * @param n position where the bit in {@code this} has to be flipped.
+   * @return {@code this ^ 2^n}.
+   * @throws ArithmeticException if {@code n < 0}.
+   */
+  def flipBit(n: Int): BigInteger = {
+    if (n < 0) {
+      throw new ArithmeticException("Negative bit address")
+    }
+    BitLevel.flipBit(this, n)
+  }
+
+
+  /**
+   * Returns this {@code BigInteger} as an float value. If {@code this} is too
+   * big to be represented as an float, then {@code Float.POSITIVE_INFINITY} or
+   * {@code Float.NEGATIVE_INFINITY} is returned. Note, that not all integers x
+   * in the range [-Float.MAX_VALUE, Float.MAX_VALUE] can be represented as a
+   * float. The float representation has a mantissa of length 24. For example,
+   * 2^24+1 = 16777217 is returned as float 16777216.0.
+   *
+   * @return this {@code BigInteger} as a float value.
+   */
+  override def floatValue(): Float = {
+    java.lang.Float.parseFloat(this.toString)
+  }
+
+  /**
+   * Returns a new {@code BigInteger} whose value is greatest common divisor of
+   * {@code this} and {@code val}. If {@code this==0} and {@code val==0} then
+   * zero is returned, otherwise the result is positive.
+   *
+   * @param bi value with which the greatest common divisor is computed.
+   * @return {@code gcd(this, val)}.
+   * @throws NullPointerException if {@code val == null}.
+   */
+  def gcd(bi: BigInteger): BigInteger = {
+    val val1 = this.abs()
+    val val2 = bi.abs()
+    // To avoid a possible division by zero
+    if (val1.signum() == 0) {
+      val2
+    } else if (val2.signum() == 0) {
+      val1
+    } else if (((val1.numberLength == 1) || ((val1.numberLength == 2) && (val1.digits(1) > 0))) &&
+        (val2.numberLength == 1 || (val2.numberLength == 2 && val2.digits(1) > 0))) {
+      // Optimization for small operands
+      // (op2.bitLength() < 64) and (op1.bitLength() < 64)
+      BigInteger.valueOf(Division.gcdBinary(val1.longValue(), val2.longValue()))
```

Comment:
My point is that I think it doesn't make sense for `gcdBinary(Long, Long)` to exist at all. There should be a `gcdBinary(Int, Int)` instead. "Optimizing" for the `Long` case is unlikely to be an optimization, since `Long`s are basically implemented with a `BigInteger` kind of strategy, only with a fixed number of underlying `Int`s.